test/e2e/providers/azure/azure.go

```diff
@@ -0,0 +1,194 @@
+)
+
+// Provider is a provider struct for testing Azure
+type Provider struct {
+	oc     *clusterinfo.OpenShift
+	vmSize string
+}
+
+// New returns a new Azure provider struct with the give client set and ssh key pair
+func New(clientset *clusterinfo.OpenShift, hasCustomVXLANPort bool) (*Provider, error) {
+	if hasCustomVXLANPort == true {
+		return nil, fmt.Errorf("custom VXLAN port is not supported on current Azure image")
+	}
+
+	return &Provider{
+		oc:     clientset,
+		vmSize: defaultVMSize,
+	}, nil
+}
+
+func newAzureMachineProviderSpec(clusterID string, status *config.PlatformStatus, location, zone, vmSize string) (*azureprovider.AzureMachineProviderSpec, error) {
+	if clusterID == "" {
+		return nil, fmt.Errorf("clusterID is empty")
+	}
+	if status == nil || status == (&config.PlatformStatus{}) {
+		return nil, fmt.Errorf("platform status is nil")
+	}
+	if status.Azure == nil || status.Azure == (&config.AzurePlatformStatus{}) {
+		return nil, fmt.Errorf("azure platform status is nil")
+	}
+	if status.Azure.NetworkResourceGroupName == "" {
+		return nil, fmt.Errorf("azure network resource group name is empty")
+	}
+	rg := status.Azure.ResourceGroupName
+	netrg := status.Azure.NetworkResourceGroupName
+
+	return &azureprovider.AzureMachineProviderSpec{
+		TypeMeta: meta.TypeMeta{
+			APIVersion: "azureproviderconfig.openshift.io/v1beta1",
+			Kind:       "AzureMachineProviderSpec",
+		},
+		UserDataSecret: &core.SecretReference{
+			Name:      clusterinfo.UserDataSecretName,
+			Namespace: clusterinfo.MachineAPINamespace,
+		},
+		CredentialsSecret: &core.SecretReference{
+			Name:      defaultCredentialsSecretName,
+			Namespace: clusterinfo.MachineAPINamespace,
+		},
+		Location: location,
+		Zone:     &zone,
+		VMSize:   vmSize,
+		Image: azureprovider.Image{
+			Publisher: defaultImagePublisher,
+			Offer:     defaultImageOffer,
+			SKU:       defaultImageSKU,
+			Version:   defaultImageVersion,
+		},
+		OSDisk: azureprovider.OSDisk{
+			OSType:     "Windows",
+			DiskSizeGB: defaultOSDiskSizeGB,
+			ManagedDisk: azureprovider.ManagedDisk{
+				StorageAccountType: defaultStorageAccountType,
+			},
+		},
+		PublicIP:             false,
+		Subnet:               fmt.Sprintf("%s-worker-subnet", clusterID),
+		ManagedIdentity:      fmt.Sprintf("%s-identity", clusterID),
+		Vnet:                 fmt.Sprintf("%s-vnet", clusterID),
+		ResourceGroup:        rg,
+		NetworkResourceGroup: netrg,
+	}, nil
+}
+
+// GenerateMachineSet generates the machineset object which is aws provider specific
+func (p *Provider) GenerateMachineSet(withWindowsLabel bool, replicas int32) (*mapi.MachineSet, error) {
+	clusterID, err := p.oc.GetInfrastructureID()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get cluster id: %v", err)
+	}
+	platformStatus, err := p.oc.GetCloudProvider()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get azure platform status: %v", err)
+	}
+
+	// Inspect master-0 to get Azure Location and Zone
```

Comment:
Would it be better to use a Linux worker here?

Reply:
My reasoning here is that master-0 _always_ exists.
While we do have the Linux workers on CI, it is technically possible to install a cluster without workers (even though that's definitely an edge case).
If you prefer taking the location and zone from a worker, I can change it and it'll still work on CI.

Reply:
I am fine leaving it as is.